base/system/winlogon/sas.c

```diff
@@ -365,15 +365,19 @@ PlayLogonSound(
         CloseHandle(hThread);
 }
 
+/* NOTE: Logoff & shutdown sound play synchronously. */
 static
 VOID
-PlayLogoffSound(
-    _In_ PWLSESSION Session)
+PlayLogoffShutdownSound(
+    _In_ PWLSESSION Session,
+    _In_ BOOL bShutdown)
 {
     if (!ImpersonateLoggedOnUser(Session->UserToken))
         return;
 
-    PlaySoundRoutine(L"WindowsLogoff", FALSE, SND_ALIAS | SND_NODEFAULT);
+    PlaySoundRoutine(bShutdown ? L"SystemExit" : L"WindowsLogoff",
```

Comment:
On Windows, the `WindowsLogon` and `WindowsLogoff` sounds are played by explorer.exe, while the "SystemStart" and "SystemExit" sounds are indeed played by Winlogon.
(The remark about explorer.exe may be verified by e.g. changing which shell gets started by default. Change the `REG_SZ` value named `Shell` in the `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon` registry key, to something different, for example, `cmd.exe`.)

Reply:
Some extra info: Windows' winlogon uses, for systemstart and exit sounds:
- `PlaySoundW` using `SND_ALIAS_SYSTEMSTART` (or ...EXIT) sound alias ID (with `SND_ALIAS_ID | SND_NODEFAULT`);
- before calling this function, it checks whether the system warning beeper is enabled, by calling `SystemParametersInfoW` with `SPI_GETBEEP` parameter. In case it is, it'll play the sound; otherwise it won't.

Reply:
WinXP and Win7 use different design.

WinXP:

- WindowsLogoff = switch user (to profiles page)
- WindowsLogon = switch user (back to desktop)
- SystemStart = user logon
- SystemExit = user logoff
- WindowsLogon & WindowsLogoff sounds require explorer.exe
- SystemStart & SystemExit sounds don't require explorer.exe

Win7:

- WindowsLogon = user logon
- WindowsLogoff = user logoff
- SystemStart = system startup
- SystemExit = system shutdown
- Switch user has no sound
- All sound events require explorer.exe

Reply:
Which design do you think is the best for ReactOS?

This PR likes Win7 one, but implemented in winlogon.

Reply:
Maybe some kind of discussion, then a vote can be done in mattermost (for example).
Here are some pros/cons I find:
- Pro for having the sounds in winlogon: this allows the sounds to be played even if the startup shell (e.g. explorer.exe) is changed to another app, e.g. cmd.exe.
- Con: "extra" complexity in winlogon.

I wonder what sounds Windows 2000 had. Do you know whether it supported all those (SystemStart/Exit and WindowsLogon/Logoff)? or only the SystemStart/Exit ones?

Reply:
> I wonder what sounds Windows 2000 had. Do you know whether it supported all those (SystemStart/Exit and WindowsLogon/Logoff)? or only the SystemStart/Exit ones?

I don't have Win2000 VM.

Reply:
Windows 2000 supported sounds:
![W2000_1](https://github.com/reactos/reactos/assets/86486978/a3853f62-1b0b-4634-80b4-ece46ba9698f)
![W2000_2](https://github.com/reactos/reactos/assets/86486978/b8881573-28c6-40f4-af71-856f34ef50e1)

Reply:
I think we should move```PlaySoundRoutine``` to outside of WinLogon. Explorer will need to call the ```Start Navigation``` and ```Empty Recycle Bin``` sound events. It'd be nice to use the same method everywhere to call the sound events.

Reply:
`PlaySoundRoutine` in Winlogon still needed for `MessageBeep` sounds.